### PR TITLE
Improve performance of Widget Tests

### PR DESCRIPTION
**Pull request by jiahaog**
## Description

Disable intrinsic checks for flutter_test by default, but leaves them enabled for framework tests.

These checks are expensive, and disabling them yields a 30% speed up for the slowest 35 internal targets.

This also reenables intrinsic checks for the framework tests that are not declared with `testWidgets()`. When declaring tests with `test()`, the binding is not initialized, and the value of `checkIntrinsicSizes` is not propagated into the Rendering.

It removes the [`checkIntrinsicSizes`](https://github.com/flutter/flutter/blob/678a4e882a496f40681296b1b7339f85ada577ee/packages/flutter_test/lib/src/binding.dart#L296) property from `TestWidgetsFlutterBinding`, because this property has confusing behavior depending on when it was set. Users should use [`debugCheckIntrinsicSizes`](https://master-api.flutter.dev/flutter/rendering/debugCheckIntrinsicSizes.html) instead if they wish to turn it on.

## Tests

Add a test to verify that intrinsic checks are always enabled for the framework tests.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by jiahaog** [comment]
@hixie I'm not sure if this is a breaking change though (maybe this change should be a opt-in / opt-out?) - I patched this change and ran the global presubmit internally, and there is a test that broke (sent you an email linking the breakage). I'm not too familiar with creating custom render objects, but the breakage is coming from a custom render object that is doing some weird things for their intrinsic calculations. Let me know what you think!

**Comment by Hixie** [comment]
> I'm not sure if this is a breaking change

See https://github.com/flutter/flutter/wiki/Tree-hygiene#1-determine-if-your-change-is-a-breaking-change for how to determine what is considered a breaking change. Long story short though, if it breaks a test in google3 then yes, it's a breaking change.

**Comment by Hixie** [comment]
If this breaks something, that seems concerning, though...

cc @goderbauer

**Comment by jiahaog** [comment]
Yeah, in that case this will be a breaking change. I wonder if that custom render object is the recommended way to do things though. The following is the simplified custom render object that I'm not sure how to fix-forward:

<details>
  <summary>Example</summary>

```dart
// Some descendant of RenderWrap, not necessarily a direct one.
class _CustomRenderWrap extends RenderWrap {
  bool _superDoingLayout = false;

  // Get from children
  _RenderWithOverride get overrideableRenderChild => ... 

  @override
  double computeMinIntrinsicWidth(double height) {
    if (_superDoingLayout) return super.computeMinIntrinsicWidth();

    overrideableRenderChild.sizeOverride = Size.zero;

    // This will call `overrideableRenderChild.computeMinIntrinsicWidth` down the line.
    // Now the cached computation on `overrideableRenderChild` is polluted with the size override
    // for subsequent calls.
    final computed = super.computeMinIntrinsicWidth(height);

    overrideableRenderChild.sizeOverride = null;

    return computed;
  }


  @override
  void performLayout() {
    _superDoingLayout = true;
    super.performLayout();
    _superDoingLayout = false;

    // ...
  }
}

class _OverrideWidget extends SingleChildRenderObjectWidget {
  @override
  RenderObject createRenderObject(BuildContext context) => _RenderWithOverride();
}

class _RenderWithOverride extends RenderProxyBox {
  Size sizeOverride;

  @override
  double computeMinIntrinsicWidth(double height) => sizeOverride?.width ?? super.computeMinIntrinsicWidth(height);
}
```

</example>

**Comment by Hixie** [comment]
That setup is probably already buggy. It's hard for me to pin down exactly where it'll fail. Subclassing RenderWrap is something I would discourage; composition is better than subclassing for changing layout behaviour. Render objects that talk to each other is bug-prone because their lifecycles aren't synchronised; for example, in the code above nothing seems to ensure that sizeOverride will trigger a layout when changed, so the API is subtle at best. Returning different values from computeMinIntrinsicWidth based on what is happening in performLayout will cause caching issues, as you've discovered.

It's hard for me to advise because I don't really know what this code is trying to do, but I suspect the answer is to entirely change the approach being taken here.

**Comment by jiahaog** [comment]
Taking a closer look at the breakage, I think the whole convoluted setup to override the returned computation from the child is unnecessary. The following can be the override for `_CustomRenderWrap` instead:

```
@override
double computeMinIntrinsicWidth(double height) {
  if (_superDoingLayout) return super.computeMinIntrinsicWidth();
  return 0;
}
```

`_RenderWithOverride` and `_OverrideWidget` can be deleted as well. I patched this fix internally and the global presubmit passed, so I will send a CL out for that to simplify the code. In light of this override being not exactly appropriate, will we still need to go through the breaking change process for this?

**Comment by jiahaog** [comment]
Spoke to the original author of the internal breakage and resolved it by moving them away from the separate render object. @hixie can you take a look again before I get started on the breaking change process?

**Comment by Hixie** [comment]
The point of not caching is that we want to make sure that the same result is returned every time. If we decide to not test that, then why not just use the regular cache, instead of having two caches?

I'm worried that caching is going to miss the exact types of bugs that led us to disable the cache in the first place though.

**Comment by jiahaog** [comment]
> The point of not caching is that we want to make sure that the same result is returned every time. If we decide to not test that, then why not just use the regular cache, instead of having two caches?

I think this does keep the same results is returned. If the existing `_cachedIntrinsicDimensions` is used as the sole cache for both the intrinsic checks and normal layout, there will be a subtle increase in the calls to `markParentNeedsLayout` (shown below) because `_cachedIntrinsicDimensions` will no longer be empty.

https://github.com/flutter/flutter/blob/a7f5fd5360007518644597c60a1f2169eacccc2b/packages/flutter/lib/src/rendering/box.dart#L2293-L2308

As such, I opted to add a separate cache for these intrinsic checks to avoid changing the existing behavior.

> I'm worried that caching is going to miss the exact types of bugs that led us to disable the cache in the first place though.

I noticed that #4446 added `_cachedIntrinsicDimensions` and left the cache turned off just for the intrinsic size checks, but I couldn't find out why the cache was disabled for this from history. Are there any examples of bugs that turning off the cache will catch? If the cache invalidation is correct the same bugs should still be caught by the intrinsic size checks 🙂

**Comment by Hixie** [comment]
I looked into this some more. Here's the problem in a nutshell:

If we don't call markNeedsLayout on the parent when we have cached intrinsic dimensions, then the parent's cache won't be invalidated, so it will start reporting incorrect results.

If we do call markNeedsLayout on the parent when we have cached intrinsic dimensions, then the parent will layout in debug mode (when checking intrinsics) but won't in release mode, leading to bugs being hidden in tests, the worst place to hide bugs.

In fact there's a bug in the PR right now, because [we don't invalidate the cache of the parent](https://github.com/flutter/flutter/pull/70730/files#diff-f2acffc9e582820293ac5a6dfbe0f8ac16491e992d4e89e074412b8d637f3cc6R2302) when the child is marked dirty for layout and has cached debug intrinsic dimensions.

Basically there's no way to correctly implement caching for debug-only intrinsic dimension logic.

(We should add a test that checks this, even if we don't land this PR.)

**Comment by jiahaog** [comment]
Thanks for taking a closer look! I'll have to come back on this, but I sent #71539 anyway to add the tests you mentioned (which fails when this PR is patched). I'm quite new to layout though, so do let me know if the test covers the correct thing.

**Comment by jiahaog** [comment]
Would clearing the debug intrinsic cache of the parent separately help with this?

E.g. https://github.com/jiahaog/flutter/commit/3279c0bfecd386ca10f5b569f0dd1da478fe09a5 (with this PR as the base commit), which will keep the the test added in #71539 passing.

**Comment by Hixie** [comment]
Consider the case where the parent isn't a RenderBox, but still does the debug intrinsics dance with its RenderBox children.

**Comment by jiahaog** [comment]
If I understand correctly, jiahaog@3279c0b wouldn't work because if the parent is not a `RenderBox`, clearing of the debug intrinsic cache will stop there and will not propagate to the grandparent which might be a `RenderBox`.

Just offering some ideas around this here – any thoughts about the following?
- Having a protected method in `RenderObject` to propagate clearing of the debug intrinsics to parents, just like how `markNeedsLayout` does it to avoid the aforementioned problem
- Having some sort of global variable for that consolidates debug intrinsic caches instead, such as `Map<RenderBox, Map<_IntrinsicDimensionsCacheEntry, double>>`, where keys are every `RenderBox` and values are the debug intrinsic cache for a particular `RenderBox`. So when `markNeedsLayout` is called, this global variable can be quickly cleared to remove all cached debug intrinsics

**Comment by Hixie** [comment]
> * Having a protected method in `RenderObject` to propagate clearing of the debug intrinsics to parents, just like how `markNeedsLayout` does it to avoid the aforementioned problem

RenderObject can't know about this, it's specific to RenderBox (and other subclasses that might support intrinsics). The whole point of having RenderObject and RenderBox be separate is to allow RenderObject to be ignorant about this kind of thing so people can make new rendering subclasses with different semantics. :-)

(Or to put it another way: Consider what you would do if RenderBox was not part of Flutter's framework, but was in a custom package. Would we accept a patch that modified RenderObject just to support this esoteric feature of that package?)

> * Having some sort of global variable for that consolidates debug intrinsic caches instead, such as `Map<RenderBox, Map<_IntrinsicDimensionsCacheEntry, double>>`, where keys are every `RenderBox` and values are the debug intrinsic cache for a particular `RenderBox`. So when `markNeedsLayout` is called, this global variable can be quickly cleared to remove all cached debug intrinsics

I suppose that could work... I'm really reluctant to introduce global variables though, one almost always regrets doing so. Also, would this actually improve performance? If we blow away the entire cache any time markNeedsLayout is called, there's not going to be much time that the cache actually does anything useful. It might even make things slower...

**Comment by jiahaog** [comment]
Thanks for explaining it to me! Yeah, it would make sense that the intrinsics are kept contained within the `RenderBox` implementation, it will not be appropriate to add a method to `RenderObject` just for this.

> I suppose that could work... I'm really reluctant to introduce global variables though, one almost always regrets doing so. Also, would this actually improve performance? If we blow away the entire cache any time markNeedsLayout is called, there's not going to be much time that the cache actually does anything useful. It might even make things slower...

Do you think using the global variable would be worth it though? I can't quite think of any other way to reference all the `RenderBox` in the widget tree to clear the cache properly at this time... To quantify the performance improvements, I created an internal patch using a global variable as described for the debug intrinsic cache, and ran the slowest 35 test targets. Across these targets they complete on average ~30% faster, and the runtime improvements vary to up to 70% faster for certain targets. Happy to share the data internally as well.

**Comment by Hixie** [comment]
That's a big improvement, I'm surprised. I suppose it's just that we're avoiding the O(N^2) effect because each time we walk the entire tree for intrinsics, we cache the results at each stage the whole thing becomes O(N)? I guess that makes sense...

Thinking about this some more, what would happen in the case of a RenderBox->RenderFoo adapter? e.g. consider a RenderBox whose child is a RenderFoo where the Foo protocol has some sort of intrinsics equivalent. The Box would ask the Foo for its intrinsic size, the Foo asks its child, reports the answer back to the Foo, which gives it to the Box. If RenderObject.debugCheckingIntrinsics is true, though, the child might not set up a relationship with its parent, so when it later changes, it calls markNeedsLayout on itself, but it finds it doesn't need to call markNeedsLayout on its parent. The RenderBox never gets dirtied, so the cache isn't cleared. This seems like it would then start returning bad values...

Do you still get the improvement if we just clear the cache on every frame? The binding in the rendering library already knows about box.dart so it would be reasonable for it to go and clear out the cache (you'd have to have a public method for clearing the cache, I guess, but that's ok...).

**Comment by jiahaog** [comment]
>  what would happen in the case of a RenderBox->RenderFoo adapter?

Yeah... Unfortunately this edge case wouldn't work with the global cache then :(. I tried out your suggestion to clear the cache on every frame, by doing it at the [end of `drawFrame`](https://github.com/flutter/flutter/blob/a1070e4f4c3f5717bcd2c5a0ce6e85e3cd1925d3/packages/flutter/lib/src/rendering/binding.dart#L462). This does still give the performace improvement, but it also causes some tests to fail internally with intrinsic check failures, probably because widgets might be marked dirty multiple times within a frame in between intrinsic checks.

Something else I tried was to clear the global cache instead [before every `RenderBox`s' intrinsic checks](https://github.com/flutter/flutter/blob/a1070e4f4c3f5717bcd2c5a0ce6e85e3cd1925d3/packages/flutter/lib/src/rendering/box.dart#L2214). This does not result in the failures, but the performace gains are not as significant - it averages to about 15% faster, down from 30% previously. This probably happens because there's still a bunch of wasted work recalculating intrinsics when the widgets are not dirty.

I think what we might want here is to somehow hook into all implementations of `RenderObject` and when they mark themselves as dirty, clear the global cache. There doesn't seem to be a simple way of doing this though. Even if we could add some sort of `debugAssertDoesMeetConstraintsClearGlobalCache` method to `RenderObject` API (not that it's the best idea) and place it in `RenderObject.markNeedsLayout`, implementations such as `RenderFoo` don't necessarily need to call `super.markNeedsLayout`, and there's doesn't seem to be a way of reliably knowing that `RenderFoo` is dirty.

This does seem a bit difficult to implement correctly and I can't think of anything else that might solve the edge case you pointed out at this time. I guess an option here would be to settle for the smaller performance gains, but I'd like to think a bit more about this over the weekend :)

**Comment by Hixie** [comment]
We could also make this particular test easier to turn off for entire groups of tests. @tvolkert made a thing that lets you configure the tests once, he used it to make the golden tests work. Maybe we can use that to just turn off these tests for teams where the tests are too slow? (We'd have to document how to do this somewhere on our Web site if that's what we recommend.)

**Comment by jiahaog** [comment]
I spent some more time thinking about it, and here are some ideas :)

### Option A: Add `debugMarkDirtyForAssertDoesMeetConstraints` to `RenderObject`, annotate `markNeedsLayout` with `@mustCallSuper`

Example: https://github.com/flutter/flutter/pull/72252

**~30%** faster on average for the slowest 35 internal targets.

We can introduce to `RenderObject` the concept of an accompanying cleanup method for `debugAssertDoesMeetConstraints`, named `debugMarkDirtyForAssertDoesMeetConstraints`. Descendants of `RenderObject` can implement it, and it will be called in `RenderObject.markNeedsLayout`. In this case, `RenderBox` will implement it to clear the global debug intrinsics cache. 

To make sure that clients do not override this logic in `RenderObject.markNeedsLayout`, we can add the `@mustCallSuper` annotation to `RenderObject.markNeedsLayout`. This doesn't look like a breaking change (from running internal tests).

I'm a bit on the fence about this approach, because it feels a bit more difficult to explain to users *how* to implement it for non-`RenderBox` operations. Perhaps the method can be renamed to something like `beforeMarkNeedsLayout` as a generic callback for implementators of `RenderObject` to do work.

### Option B: Add a global intrinsics cache, always clear it before intrinsic checks

Example: https://github.com/flutter/flutter/pull/72260

**~15%** faster on average for the slowest 35 internal targets.

We can also go with the smaller performance gains with the advantage of keeping the cache invalidation simpler. This will involve just wiping the entire intrinsics global cache right before the checks. 

### [Suggested] Option C: Make intrinsic checks lazy, relax checks

Example: https://github.com/flutter/flutter/pull/72258

**~30%** faster on average for the slowest 35 internal targets.

Previouly we discussed the need to use a separate cache for intrinsic checks, because we do not want to affect real layout. What if we instead simplified the intrisic checks so that we don't need a separate cache?

The eager intrinsic checks introduces additional caching keys due to the [extra constraints](https://github.com/flutter/flutter/blob/63062a64432cce03315d6b5196fda7912866eb37/packages/flutter/lib/src/rendering/box.dart#L2242-L2247) of doing intrinsic computations with `double.infinity`, and `constraints.min/max/height/width`, which might never be encountered in the lifetime of the `RenderBox`. We could instead validate just the values that are already computed as part of real layout. The example PR above probably needs a bit more work and structure, but it shows this in more detail. 

The gist of it is to make the checks lazy and use the real layout cache for the intrinsic checks. `_computeIntrinsicDimension` is augmented to do the intrinsics checks instead as and when values are returned. It checks if the returned values are not negative and finite, and attempts to check for min < max as and when the values are saved into the cache.

### Option D: Figure out some way to make it easier to turn the checks off

Turning off the intrinsic size checks makes the tests **~35%** faster on average for the slowest 35 internal targets.

(Assuming the prior art for configuring golden tests is #16859)

Currently it is possible to turn the intrinsic checks off by subclassing either of the `TestWidgetsFlutterBinding`s, and overriding the getter [`checkIntrinsicSizes`](https://master-api.flutter.dev/flutter/flutter_test/TestWidgetsFlutterBinding/checkIntrinsicSizes.html). Perhaps we can make this instance property a setter or even a top level library setter (like [`goldenFileComparator`](https://master-api.flutter.dev/flutter/flutter_test/goldenFileComparator.html)) which will make it easier for users to toggle.

This is more convenient but I feel like it might end up being buried somewhere in the documentation as a configuration though. Perhaps we can add some sort of heuristic to the tool or `package:flutter_test`, for when tests exceed a certain time to tell users of the existence of that configuration to make it more discoverable too.

### Option E: Option D + any other option listed above

Option D of making the flag easier to configure is not exclusive with the other options listed, and a solution could be a combination of those options.

---

Personally I like option C better, but I'm not sure if relaxing the checks is alright as they might still offer some value. What do you think?

**Comment by Hixie** [comment]
The problem with C is that it doesn't catch the problems we want to catch. The real goal of these asserts, at the end of the day, are to verify that the render objects we write in the framework all implement all the right intrinsic logic. Admittedly they don't catch every problem, we still have some errors that fall through the cracks, but at least they catch a bunch of them. (We never see those failures because they get caught when someone tries to implement intrinsics and gets it wrong, before they land their PR.)

A sounds like it would make our developer experience worse so i'd steer away from that. B might work, but D might just be easier. We could even default the tests off and set the flag in the framework's configuration.

**Comment by jiahaog** [comment]
Got it! Sounds like we want to explore option D then. 

> We could even default the tests off and set the flag in the framework's configuration.

Can you elaborate a bit more about what you mean by the framework configuration here?

**Comment by Hixie** [comment]
> Can you elaborate a bit more about what you mean by the framework configuration here?

https://github.com/flutter/flutter/blob/master/packages/flutter/test/flutter_test_config.dart and related files.

**Comment by tvolkert** [comment]
See the package docs in https://api.flutter.dev/flutter/flutter_test/flutter_test-library.html

File: packages/flutter/lib/src/rendering/debug.dart
Comment (jiahaog) [inline packages/flutter/lib/src/rendering/debug.dart:286]: I'm not really sure about this change here. Currently, `debugCheckIntrinsicSizesOverride` depends on the value set in the `checkIntrinsicSizes` instance variable of `TestWidgetsFlutterBinding`. Even if we make `TestWidgetsFlutterBinding.checkIntrinsicSizes` a setter, we will still need to initialize the binding before changing the value in `flutter_test_config.dart`. Some tests in the framework do not use `TestWidgetsFlutterBinding` though, so initializing `TestWidgetsFlutterBinding` and then changing the value of `TestWidgetsFlutterBinding.checkIntrinsicSizes` would change the semantics of the tests.

If we're directing users to set `debugCheckIntrinsicSizes` from `package:flutter/rendering.dart`, then this check here should probably be removed. Another option would be to introduce a new nullable top level setter in `package:flutter_test`, and change the logic of `TestWidgetsFlutterBinding.checkIntrinsicSizes` to 

```dart
bool get checkIntrinsicSizes => newTopLevelFlutterTestcheckIntrinsicSizes ?? false;
```

We could even deprecate the `TestWidgetsFlutterBinding.checkIntrinsicSizes` to make it such that there's only one way to turn on the checks (via `debugCheckIntrinsicSizes` from `package:flutter/rendering.dart`) and keeping it simpler. This might not even be a breaking change because it's not used in Google3 and probably not by customer tests (need to verify this).

**Review by jiahaog — COMMENTED** [review]
Thanks for the pointers, I have updated this PR to disable the checks by default while leaving it enabled in the `flutter_test_config.dart` for the framework. As part of this, I also noticed that the intrinsic size checks are actually not turned on today in the framework tests, for test scripts that do not use `testWidgets`. This PR will fix that.  

On the other hand, it seems like `flutter_test_config.dart` doesn't work on web tests, so the new test I added breaks. I'll have to investigate a bit further...

**Comment by jiahaog** [comment]
Yup, `flutter_test_config.dart` is not supported on web platforms. Sending #72488 separately to fix that.

**Comment by jiahaog** [comment]
#72488 has been merged, so updated the PR and the description. Please take a look again @Hixie :)

File: packages/flutter/lib/src/rendering/debug.dart
Comment (Hixie) [reply packages/flutter/lib/src/rendering/debug.dart:286]: The purpose of these checks is to make sure that someone doesn't set the flag in a test, then forget to unset it, so that it affects all subsequent tests.

I don't mind exactly how we do that. For most other flags we just check that it's not set, but that would mean that we couldn't override it globally. We could store the value at the start of the test, then check here that it matches the value at the start of the test. That would still catch people who fiddle with the flag in a test, without preventing the use of the flag as a global override.

File: packages/flutter_test/lib/src/binding.dart
Comment (Hixie) [inline packages/flutter_test/lib/src/binding.dart:295]: is this used anymore?

File: packages/flutter/lib/src/rendering/debug.dart
Comment (jiahaog) [reply packages/flutter/lib/src/rendering/debug.dart]: > We could store the value at the start of the test, then check here that it matches the value at the start of the test

Updated `_verifyInvariants` in `TestWidgetsFlutterBinding` to do this.

File: packages/flutter_test/lib/src/binding.dart
Comment (jiahaog) [reply packages/flutter_test/lib/src/binding.dart]: This is still used when the binding is initialized in https://github.com/flutter/flutter/blob/d83134ae366e4b2e93c874f7a51260271d88c21a/packages/flutter_test/lib/src/binding.dart#L181-L185

So users have a few ways to set intrinsic sizes (in order of precedence)

1. After the binding is initialzied, set `debugCheckIntrinsicSizes` from `rendering/debug.dart`, either within a test (and resetting the value) or outside of the test
2. Override `checkIntrinsicSizes` on `TestWidgetsFlutterBinding` to a non-null value, then initialize that binding manually
3. Before the test binding is initialized, set `debugCheckIntrinsicSizes` from `rendering/debug.dart`

This is very confusing though, because of duplicated state. I'd suggest removing `checkIntrinsicSizes` from flutter test entirely, and telling users to set the global directly. It [doesn't look like](https://github.com/flutter/flutter/pull/73759) a breaking change, will that be alright?

Comment (jiahaog) [reply packages/flutter_test/lib/src/binding.dart]: I thought about it a bit more and the two ways of turning on the intrinsic size checks are really not ideal. The latest commit (https://github.com/flutter/flutter/pull/70730/commits/f56b35f1c7200a56cbef7a0e2b3b35bc56c35588) removes `checkIntrinsicSizes` from `TestWidgetsFlutterBinding` to simplify this. If that's not okay let me know and I'll revert it.

**Review by Hixie — APPROVED** [review]
LGTM

I love it.

File: packages/flutter/lib/src/rendering/box.dart
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart:1347]: we might want to put this on the docs for the computeIntrinsic* methods too, come to think of it

Comment (jiahaog) [reply packages/flutter/lib/src/rendering/box.dart:1347]: Done!

**Comment by jiahaog** [comment]
Thanks for the review and all the help! I'll look into manually turning on these checks for internal tests that involve widgets with custom intrinsic overrides.

**Comment by Hixie** [comment]
Thank you for sticking with it even through my faffing around and pedantry!

**Pull request merged**
